generators/chipyard/src/main/scala/HarnessBinders.scala

```diff
@@ -0,0 +1,279 @@
+        t match {
+          case system: T => up(HarnessBinders, site)(tag.runtimeClass.toString)(system, th, ports) ++ fn(system, th, ports)
+          case _ => Nil
+        }
+      })
+  )
+})
+
+class WithGPIOTiedOff extends OverrideHarnessBinder({
+  (system: HasPeripheryGPIOModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    ports.map { case p: GPIOPortIO => p <> AnalogConst(0) }
+    Nil
+  }
+})
+
+// DOC include start: WithUARTAdapter
+class WithUARTAdapter extends OverrideHarnessBinder({
+  (system: HasPeripheryUARTModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    UARTAdapter.connect(ports.map(_.asInstanceOf[UARTPortIO]))(system.p)
+    Nil
+  }
+})
+// DOC include end: WithUARTAdapter
+
+class WithSimSPIFlashModel(rdOnly: Boolean = true) extends OverrideHarnessBinder({
+  (system: HasPeripherySPIFlashModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    SimSPIFlashModel.connect(ports.map(_.asInstanceOf[SPIChipIO]), th.harnessReset, rdOnly)(system.p)
+    Nil
+  }
+})
+
+class WithSimBlockDevice extends OverrideHarnessBinder({
+  (system: CanHavePeripheryBlockDeviceModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val clock = WireInit(false.B.asClock)
+    ports.map {
+      case p: BlockDeviceIO => SimBlockDevice.connect(clock, th.harnessReset.asBool, Some(p))(system.p)
+      case c: Clock => clock := c
+    }
+    Nil
+  }
+})
+
+class WithBlockDeviceModel extends OverrideHarnessBinder({
+  (system: CanHavePeripheryBlockDeviceModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val clock = WireInit(false.B.asClock)
+    ports.map {
+      case p: BlockDeviceIO => withClockAndReset(clock, th.harnessReset) { BlockDeviceModel.connect(Some(p))(system.p) }
+      case c: Clock => clock := c
+    }
+    Nil
+  }
+})
+
+class WithLoopbackNIC extends OverrideHarnessBinder({
+  (system: CanHavePeripheryIceNICModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val clock = WireInit(false.B.asClock)
+    ports.map {
+      case p: NICIOvonly => withClockAndReset(clock, th.harnessReset) {
+        NicLoopback.connect(Some(p), system.p(NICKey))
+      }
+      case c: Clock => clock := c
+    }
+    Nil
+  }
+})
+
+class WithSimNetwork extends OverrideHarnessBinder({
+  (system: CanHavePeripheryIceNICModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val clock = WireInit(false.B.asClock)
+    ports.map {
+      case p: NICIOvonly => SimNetwork.connect(Some(p), clock, th.harnessReset.asBool)
+      case c: Clock => clock := c
+    }
+    Nil
+  }
+})
+
+class WithSimAXIMem extends OverrideHarnessBinder({
+  (system: CanHaveMasterAXI4MemPort, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val p: Parameters = chipyard.iobinders.GetSystemParameters(system)
+    val clock = WireInit(false.B.asClock)
+    ports.filter(_.isInstanceOf[Clock]).map { case p: Clock => clock := p }
```

Comment:
isn't the filter redundant if you are matching on the type as well? I think restructuring it as in my above nit would make the duplication clear.